docs/smoothquant.md

```diff
@@ -0,0 +1,337 @@
+>>>print(f'{W_q}\n{W_scale.item()}')
+>>>print(f'{X_q}\n{X_scale.item()}')
+tensor([[ 13., 127., 101.],
+        [ 64.,  84., 120.]])
+0.0059755356051027775
+tensor([[ 83.,  89., 119.,  85.],
+        [ 57.,  50.,  11.,  21.],
+        [ 91.,   1.,  11., 127.]])
+0.006533813662827015
+
+>>>Y_q = torch.matmul(W_q, X_q)
+>>>Y_q
+tensor([[17509.,  7608.,  4055., 16599.],
+        [21020., 10016.,  9860., 22444.]])
+>>>Y_dq = dequantize(Y, W_scale * X_scale)
+>>>Y_dq
+tensor([[0.6836, 0.2970, 0.1583, 0.6481],
+        [0.8207, 0.3911, 0.3850, 0.8763]])
+```
+
+#### Per-channel limitation
+
+Though per-channel quantization could bring lower quantization error, we could not apply it for activations due to the difficulty of the dequantization. We would prove it in the following image and the zero point of quantization would be ignored for simplicity.
+
+The image on the left presents a normal linear forward  with 1x2 input $x$ and 2x2 weight $w$. The results $y$ could be easily obtained by simple mathematics. In the middle image, we apply per-tensor quantization for activations and per-channel quantization for weights; the results after quantization that are denoted by $y_1$ and $y_2$, could be easily dequantized to the float results $y_{fp1}$ and $y_{fp2}$ by per channel scale $1.0/s_1s_x$ and $1.0/s_2s_x$. However, after applying per-channel quantization for activation (right image), we could not dequantize the  $y_1$ and  $y_2$ to float results.
+
+<div align="center">
+    <img src="./imgs/sq_pc.png"/>
+</div>
+
+
+## SmoothQuant and Our Enhancement
+
+### SmoothQuant
+
+In the previous subsection, we have explained why per-channel quantization could not be applied for activation, even though it could lead to lower quantization loss. However, the quantization error loss of activation plays an important role in the accuracy loss of model quantization[^2][^3][^4]. 
+
+
+
+To reduce the quantization loss of activations, lots of methods have been proposed. In the following, we briefly introduce SPIQ[^2], Outlier Suppression[^3] and Smoothquant[^4]. All these three methods share a similar idea to migrate the difficulty from activation quantization to weight quantization but differ in how much difficulty to be transferred.
+
+
+So **the first question is how to migrate the difficulty from activation to weights?** The solution is straightforward, that is to convert the network to an output equivalent network that is presented in the image below and apply quantization to this equivalent network. The intuition is that each channel of activation could be scaled to make it more quantization-friendly, similar to a fake per-channel activation quantization.
+
+<div align="center">
+    <img src="./imgs/sq_convert.png"/>
+</div>
+
+
+Please note that this conversion will make the quantization of weights more difficult, because the scales attached to weights shown above are per-input-channel, while quantization of weights is per-output-channel or per-tensor.

```

Comment:
it's better to add the real case, ex: 
fig 4 in https://www.arxiv-vanity.com/papers/2211.10438/